soc/arm/atmel_sam0/common/soc_samr3x_radio_off.c

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2021 Argentum Systems Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/init.h>
+
+/* When the radio is not in use, it's important that #CS is set high, to avoid
+ * unexpected behavior and increased current consumption... see Chapter 10 of
+ * DS70005356C. We also hold the radio in reset.
+ */
+static int radio_off_setup(void)
+{
+	int ret;
+	const struct gpio_dt_spec cs = GPIO_DT_SPEC_GET(DT_NODELABEL(sercom4), cs_gpios);
+	const struct gpio_dt_spec reset = GPIO_DT_SPEC_GET(DT_NODELABEL(lora), reset_gpios);
+
+	if (!gpio_is_ready(&cs) || !gpio_is_ready(&reset)) {
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&cs, GPIO_OUTPUT_INACTIVE);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gpio_pin_configure_dt(&reset, GPIO_OUTPUT_ACTIVE);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+SYS_INIT(radio_off_setup, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```

Comment:
My original project actually operates the radio in FSK mode, with an out-of-tree driver...

`CONFIG_APPLICATION_INIT_PRIORITY` defaults to `90`, making this `APPLICATION-90`, which I think is _far_ too late to be applying this fallback / default... How would you feel about something closer to `POST_KERNEL-0`?

Reply:
> Sounds like your driver should then be responsible for applying these settings correctly, and not via a SYS_INIT call.

It does set things up, and it uses a `DEVICE_DT_INST_DEFINE(..., POST_KERNEL, 95, ...)`, which boils down to the same thing in terms of boot sequencing.

The issue is that this initialization is gated by `NOT CONFIG_LORA`, which will run this code in my situation. For what is effectively a very lightweight _driver_, I don't see any good reason for this to occur at the end of _application_ initialization, as it'll disrupt any drivers that do happen to initialize before it - an edge-case perhaps, but still desirable.

The new check conditions (i.e: `SAMRx and !LORA`) are probably more appropriate than the old (i.e: `SERCOM4 = SPI`), and I wouldn't want to suggest placing some other gate on this (e.g: `CONFIG_SAMR3X_RADIO_NOINIT`)

See also below regarding putting these signals into a known state ASAP.

> In fact, I have not found anything to what the comment says in the referenced document:
> What is the "unexpected behavior"? Is the current consumption increase documented in DS70005356C?

The datasheet states:

![image](https://user-images.githubusercontent.com/36298367/234698458-195783a6-453b-483c-8c20-82e88cc830c1.png)

Fundamentally, there are no bias resistors on these signals in the SiP, which means that leaving them Hi-Z will allow them to float... Thus the CPU_OFF sleep mode cannot be used safely, and these signals cannot be left in their default / Hi-Z state for any duration. Metastability can significantly increase current consumption, and while the unexpected behavior isn't detailed, it could potentially include **_anything_** the radio could do on purpose or by mistake... Setting these signals to a known state _as soon as possible_ is the ideal, even if the radio later gets initialized by a driver (in-tree or out-of-tree) - the old `soc_pinconf_init()` would run very early, even if the Zephyr LoRa driver came along later.

As a side note, during development I've actually found that sleep power can be reduced beyond holding the part in reset by commanding it into a sleep state, but that feels like an excessively large driver for what is not likely a production situation.

Reply:
Thanks for all the details, I think i understand the problem better now. I've changed the approach in the latest revision by adding a new option:

```
config SOC_ATMEL_SAMR3X_RADIO_OFF_SETUP
	bool "Configure LoRa radio pins if not in use"
	default y if !LORA
	depends on SOC_SERIES_SAMR34 || SOC_SERIES_SAMR35
	help
	  As detailed in DS70005356C, LoRa radio SPI pins do not have pull-ups,
	  so when the radio is not in use, it's important that CS is kept high,
	  to avoid unexpected behavior and increased current consumption. To
	  further reduce power consumption, radio can be kept in reset state by
	  keeping nRST pin low. When enabling this option, both CS and nRST will
	  be configured high and low, respectively.
```

In OOT boards/applications, you can either `CONFIG_SOC_ATMEL_SAMR3X_RADIO_OFF_SETUP=n`, or, in board `Kconfig.defconfig`,

```
config SOC_ATMEL_SAMR3X_RADIO_OFF_SETUP
	default y if !MY_OOT_DRIVER_CLASS
```

I think this solution is what gives the most flexibility.

Reply:
I think the ideal is that this _always_ happens _ASAP_ for SAMR34 / SAMR35 parts, regardless of a driver potentially stepping in later... perhaps `default y` or even removing `SOC_ATMEL_SAMR3X_RADIO_OFF_SETUP` completely? (sorry)

(I was wrong in my previous comment about this init _always_ happening, but that would probably be preferred)

Reply:
> is it really necessary to run this code?
> there any real chance of the radio doing something weird until the driver is initialized?

I'd say that's really difficult to say, until someone spends multiple hours chasing down a weird / temperamental startup issue... It's certainly the "_safe_" option.

>  I imagine it's uncommon to not use the radio on SAMR3x
> So, I assume it's useful in the context of Zephyr, where you have many generic samples that don't use the radio.

Yes, agreed - this is definitely **required** for situations where the radio isn't used, including early bring-up too.

Reply:
> > is it really necessary to run this code?
> > Is there any real chance of the radio doing something weird until the driver is initialized?
> 
> That's really difficult to say, until someone spends multiple hours chasing down a weird / temperamental startup issue... It's certainly the "_safe_" option - and it's a very short function.

Do you know how Atmel official samples handle this?

Reply:
Sorry I never got back to you on this one, I've been stretched thin... it's merged now, so let's leave it be, but keep it in mind for any issues that may arise down the road.